client/mainmenu/CHighScoreScreen.cpp

```diff
@@ -0,0 +1,401 @@
+		}
+
+		if(curData["points"].Integer() > 0)
+			images.push_back(std::make_shared<CAnimImage>(AnimationPath::builtin("CPRSMALL"), (*CGI->creh)[HighScoreCalculation::getCreatureForPoints(curData["points"].Integer(), highscorepage == HighScorePage::CAMPAIGN)]->getIconIndex(), 0, 670, y - 15 + i * 50));
+	}
+}
+
+void CHighScoreScreen::buttonCampaginClick()
+{
+	highscorepage = HighScorePage::CAMPAIGN;
+	addHighScores();
+	addButtons();
+	redraw();
+}
+
+void CHighScoreScreen::buttonStandardClick()
+{
+	OBJ_CONSTRUCTION_CAPTURING_ALL_NO_DISPOSE;
+	highscorepage = HighScorePage::SCENARIO;
+	addHighScores();
+	addButtons();
+	redraw();
+}
+
+void CHighScoreScreen::buttonResetClick()
+{
+	CInfoWindow::showYesNoDialog(
+		CGI->generaltexth->allTexts[666],
+		{},
+		[this]()
+		{
+			Settings entry = persistentStorage.write["highscore"];
+			entry->clear();
+			addHighScores();
+			addButtons();
+			redraw();
+		},
+		0
+	);
+}
+
+void CHighScoreScreen::buttonExitClick()
+{
+	close();
+}
+
+CHighScoreInputScreen::CHighScoreInputScreen(bool won, HighScoreCalculation calc)
+	: CWindowObject(BORDERED), won(won), calc(calc)
+{
+	addUsedEvents(LCLICK | KEYBOARD);
+
+	OBJ_CONSTRUCTION_CAPTURING_ALL_NO_DISPOSE;
+	pos = center(Rect(0, 0, 800, 600));
+	updateShadow();
+
+	background = std::make_shared<TransparentFilledRectangle>(Rect(0, 0, pos.w, pos.h), Colors::BLACK);
+
+	if(won)
+	{
+		int border = 100;
+		int textareaW = ((pos.w - 2 * border) / 4);
+		std::vector<std::string> t = { "438", "439", "440", "441", "676" };
+		for (int i = 0; i < 5; i++)
+			texts.push_back(std::make_shared<CMultiLineLabel>(Rect(textareaW * i + border - (textareaW / 2), 450, textareaW, 100), FONT_HIGH_SCORE, ETextAlignment::TOPCENTER, Colors::WHITE, CGI->generaltexth->translate("core.genrltxt." + t[i])));
+
+		std::string creatureName = (calc.calculate().cheater) ? CGI->generaltexth->translate("core.genrltxt.260") : (*CGI->creh)[HighScoreCalculation::getCreatureForPoints(calc.calculate().total, calc.isCampaign)]->getNameSingularTranslated();
+		t = { std::to_string(calc.calculate().sumDays), std::to_string(calc.calculate().basic), CGI->generaltexth->translate("core.arraytxt." + std::to_string((142 + calc.parameters[0].difficulty))), std::to_string(calc.calculate().total), creatureName };
+		for (int i = 0; i < 5; i++)
+			texts.push_back(std::make_shared<CMultiLineLabel>(Rect(textareaW * i + border - (textareaW / 2), 530, textareaW, 100), FONT_HIGH_SCORE, ETextAlignment::TOPCENTER, Colors::WHITE, t[i]));
+ 
+		CCS->musich->playMusic(AudioPath::builtin("music/Win Scenario"), true, true);
+	}
+	else
+		CCS->musich->playMusic(AudioPath::builtin("music/UltimateLose"), false, true);
+
+	video = won ? "HSANIM.SMK" : "LOSEGAME.SMK";
+}
+
+int CHighScoreInputScreen::addEntry(std::string text) {
+	for (int i = 0; i < 11; i++)
+	{
+		if(calc.calculate().cheater)
+			i = 10;
+
+		JsonNode node = persistentStorage["highscore"][calc.isCampaign ? "campaign" : "scenario"][std::to_string(i)];
```

Comment:
Perhaps store results as vector?
```suggestion
		JsonNode node = persistentStorage["highscore"][calc.isCampaign ? "campaign" : "scenario"][i];
```

And then for moving entries down - sort using custom functor? E.g.

```cpp
auto sortFunctor = [](const JsonNode & left, const JsonNode & right)
{
    return left["points"].Integer() < right["points"].Integer();
};
boost::range::sort(node.Vector(), sortFunctor);
```

Reply:
We cannot access  JSON with `[i]`, I already tried it. :(

Reply:
Will recheck.
There is definitely Vector() member that gives you vector of JsonNode's.
Not sure about operator []